docs/v1-proposal.md

```diff
@@ -0,0 +1,166 @@
+# v1 client changes proposal - storage refactoring
+
+With v1 approaching, I want to take a moment to look at changes we want to make
+to the existing client libraries to better set us up for long term maintenance.
+
+We already know that we want to reduce the external library surface of chains.
+But to do this, we need to define better interfaces between components that we
+expect external clients to use.
+
+Today, I think a lot of the codebase's complexity has come from a few places:
+
+1. Storage libraries
+
+   Each storage type needs different pieces of data - i.e. Grafeas and OCI need
+   to distinguish image signatures and attestation formats, and some clients
+   need the original object to extract out information like GVKs, names,
+   namespaces, etc. This has led to a organic growth of the chains libraries to
+   pass the different types of data around, and a lot of typecasting and other
+   generic object tricks.
+
+   Looking at the storage interfaces, I think this data roughly boils down to:
+
+   - The original Tekton object
+   - The formatted data object
+   - The signed payload + signature (with optional cert information)
+
+   Unlike when chains first started, we now have another useful tool available
+   to us: generics. I think we can use this to create clearer interfaces.
+
+2. Dependence on the config package
+
+   tkn depends on the chains server config, but it probably shouldn't. We should
+   aim to have better ways to initialize clients for others to use.
+
+Good news, I don't think we're far off, but we should make some changes
+
+## Signables
+
+At it's core, Chains is basically an ETL pipeline. We Extract artifacts from run
+objects, Transform and sign them, then Load them into storage.
```

Comment:
This is a good explanation. It would be great to show how they map to the `Signable`, `Payloader`, `Signer` and `Storer`  interfaces below.

I think you mean something like this: We Extract artifacts from run objects (`Signable`), Transform (`Payloader`) and sign (`Signer`) them, then Load them into storage (`Storer`). If so, that makes perfect sense to me. 

Maybe in the future, we may want to rename these interfaces. `Signer` and `Storer` are clear. `Signable` is confusing because it has nothing to do with signing - maybe `Extractor`? `Payloader` is a bit ambiguous - `Transformer` then `Transformer.Transform`? I'm nitpicking. Just trying to make this easier for newcomers to understand.

Reply:
Yeah I agree. I'm somewhat biasing on existing interface names at the moment. Happy to rename these (though probably in a separate PR).